### PR TITLE
Add VALIJSON_USE_EXCEPTIONS mode 

### PR DESCRIPTION
**Pull request by baylesj**
This patch adds support for a new Cmake option,
VALIJSON_USE_EXCEPTIONS. If specified and set to `0`, Valijson will
disable all exception throwing, add the `-fno-exceptions` compiler flag,
and print to std::err and abort where exceptions would have been thrown.

NOTE: to set the value of a CMake option, the easiest way is to modify
the appropriate source line in the <build folder>/CMakeCache.txt file.

Bug: #106

### CODE REVIEW

**Comment by baylesj** [comment]
I have a few things to ask about this patch:

1. Do you like the VALIJSON_USE_EXCEPTIONS flag option? Wanted to try this first since it looks to me like the easiest way to get the behavior we want. The approach is based somewhat on what we do in JsonCpp.

2. Is there a preferred formatting tool for this library? e.g. `clang-format`?

**Comment by tristanpenman** [comment]
I like the look of this overall. Good idea using std::abort to keep things simple. I noticed that the build is failing on Travis CI because the version being used is ancient (1.55) and the `boost/assert/source_location.hpp` header cannot be found. The builds currently use Ubuntu 14.04 (trusty) so that may be a bit of a pain to resolve.

As far as long term plans go (i.e. 1.0), my intention is to restructure things so that errors that previously resulted in an exception being thrown are properly progated back up to the caller, with appropriate context. This is currently a weakness in the library, as the generic error messages produced by the schema parser make it difficult to locate the actual errors in a schema document.

Formatting has all be pretty ad-hoc so far. I made a bit of an effort to normalise the style late last year, but haven't gone as far as applying clang-format or any other tools.

**Comment by tristanpenman** [comment]
If it turns out to be painful to fix up the build failures on Travis CI, I wouldn't be against excluding the boost property_tree related stuff from the build by default. Then we could just break that out into a separate issue, so as to not block this PR being merged.

File: include/valijson/utils/nlohmann_json_utils.hpp [hunk 1]
```diff
@@ -19,13 +20,17 @@ inline bool loadDocument(const std::string &path, nlohmann::json &document)
...
     try {
         document = nlohmann::json::parse(file);
     } catch (std::invalid_argument const& exception) {
         std::cerr << "nlohmann::json failed to parse the document\n"
             << "Parse error:" << exception.what() << "\n";
         return false;
     }
+#else
+    document = nlohmann::json::parse(file, nullptr, false);
+#endif
```
Comment (tristanpenman) [inline include/valijson/utils/nlohmann_json_utils.hpp:33]: Could be worth checking whether the result is discarded here, and returning false if so.

Comment (baylesj) [reply include/valijson/utils/nlohmann_json_utils.hpp:33]: Great idea, I'll add it.

**Comment by baylesj** [comment]
> If it turns out to be painful to fix up the build failures on Travis CI, I wouldn't be against excluding the boost property_tree related stuff from the build by default. Then we could just break that out into a separate issue, so as to not block this PR being merged.

Doesn't look too crazy, at least the current error. Rerunning it and I'll keep fixing it until it works :).

**Comment by baylesj** [comment]
As far as formatting goes, once we get this landed I'm happy to look at adding clang-format or similar.

**Comment by tristanpenman** [comment]
Yeah, sounds good. If the changes aren't too disruptive it should be okay to merge right away. Though if it's going to change a lot of files, it will probably be better to wait until I push a 1.0-dev branch later this week.

**Comment by baylesj** [comment]
Hmmm... looks like Travis isn't running on the current revision. Can you check what's going on with it? Might just be tired of me breaking it for now.

**Comment by baylesj** [comment]
NVM, found it in the travis-ci history. Still broken!

**Comment by baylesj** [comment]
Fixed the build issues and Travis seems happy now:

https://travis-ci.org/github/tristanpenman/valijson/builds/742360538

**Comment by tristanpenman** [comment]
Nice work! This is all looks good to me, so I'll go ahead and merge it.

**Pull request merged**